### ISSUE TITLE
Sampling Fails on Unevaluated `Contains` Expressions

### ISSUE THREAD

**Issue by craymichael** [sympy/sympy#20563]
Minimum code to reproduce:

```python
import sympy as S
from sympy import stats
x = S.Symbol('x')
interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
U = stats.Uniform('U', -1, 1)
next(stats.sample(U, interval.contains(U)))
```

This gives the traceback:
```
Traceback (most recent call last):
  File "/home/zach/.local/share/virtualenvs/PostHocExplainerEvaluation-d73ZRPVE/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-bb4de631a81f>", line 6, in <module>
    next(stats.sample(U, interval.contains(U)))
  File "/home/zach/.local/share/virtualenvs/PostHocExplainerEvaluation-d73ZRPVE/lib/python3.8/site-packages/sympy/stats/rv.py", line 1220, in return_generator
    gd = given_fn(*args)
  File "<lambdifygenerated-10>", line 4, in _lambdifygenerated
    logical_and.reduce((greater(U, 0),Contains(U, Interval(-1, 1)))))
NameError: name 'Contains' is not defined
```

To mitigate, manually specifying the `sample_iter_subs` function works - except in version 1.7. In version 1.6 the Contains expression would evaluate, but this appears to have regressed in 1.7.

**Comment by oscarbenjamin** [comment]
> In version 1.6 the Contains expression would evaluate

Which Contains expression would evaluate in 1.6?

I get the same exception when testing your code sample in 1.6.2:
```python
In [3]: import sympy                                                                                                              

In [4]: sympy.__version__                                                                                                         
Out[4]: '1.6.2'

In [5]: import sympy as S 
   ...: from sympy import stats 
   ...: x = S.Symbol('x') 
   ...: interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1)) 
   ...: U = stats.Uniform('U', -1, 1) 
   ...: next(stats.sample(U, interval.contains(U)))                                                                               
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-5-bb4de631a81f> in <module>
      4 interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
      5 U = stats.Uniform('U', -1, 1)
----> 6 next(stats.sample(U, interval.contains(U)))

~/current/sympy/sympy/sympy/stats/rv.py in sample(expr, condition, size, **kwargs)
   1038 
   1039     """
-> 1040     return next(sample_iter(expr, condition, size=size, numsamples=1))
   1041 
   1042 

~/current/sympy/sympy/sympy/stats/rv.py in sample_iter(expr, condition, size, numsamples, **kwargs)
   1077     # lambdify is much faster but not as robust
   1078     try:
-> 1079         return sample_iter_lambdify(expr, condition, size=size, numsamples=numsamples, **kwargs)
   1080     # use subs when lambdify fails
   1081     except TypeError:

~/current/sympy/sympy/sympy/stats/rv.py in sample_iter_lambdify(expr, condition, size, numsamples, **kwargs)
   1154     fn(*args)
   1155     if condition:
-> 1156         given_fn(*args)
   1157 
   1158     def return_generator():

<lambdifygenerated-2> in _lambdifygenerated(U)
      2     return (  # Not supported in Python with SciPy:
      3   # Contains
----> 4 logical_and.reduce((greater(U, 0),Contains(U, Interval(-1, 1)))))

NameError: name 'Contains' is not defined
```

**Comment by oscarbenjamin** [comment]
@Smit-create @czgdp1807 is this a regression that needs to be fixed in 1.7.1?

**Comment by czgdp1807** [comment]
Code,
```python
import sympy as S
print(S.__version__)
from sympy import stats
x = S.Symbol('x')
interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
U = stats.Uniform('U', -1, 1)
next(stats.sample(U, interval.contains(U)))
```

Output,
```python
1.6

---------------------------------------------------------------------------

NameError                                 Traceback (most recent call last)

<ipython-input-7-00a8c0ee57cf> in <module>()
      5 interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
      6 U = stats.Uniform('U', -1, 1)
----> 7 next(stats.sample(U, interval.contains(U)))

2 frames

/usr/local/lib/python3.6/dist-packages/sympy/stats/rv.py in sample_iter_lambdify(expr, condition, size, numsamples, **kwargs)
   1154     fn(*args)
   1155     if condition:
-> 1156         given_fn(*args)
   1157 
   1158     def return_generator():

<lambdifygenerated-4> in _lambdifygenerated(U)
      2     return (  # Not supported in Python with SciPy:
      3   # Contains
----> 4 logical_and.reduce((greater(U, 0),Contains(U, Interval(-1, 1)))))

NameError: name 'Contains' is not defined
```
Doesn't seem like a regression issue.

**Comment by czgdp1807** [comment]
Just a note on fixing this. 
The error actually arises while checking if the sampled values satisfy the given condition or not. This is done by calling lambdify on the condition. Since, lambdify generated function is unable to work with un-evaluated `Contains`, the error is thrown. There are two ways to fix this, either modify `lambdify` or change the method of ensuring that sampled values always satisfy the given condition. The former will help other modules which use lambdify and the latter will be more of a work around for `sympy.stats` only.
```python
IPython console for SymPy 1.8.dev (Python 3.6.9-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/dev


In [1]: interval = ConditionSet(x, Gt(x, 0), Interval(-1, 1))

In [2]: f = lambdify(x, interval.contains(x))

In [3]: f(0.1)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-1fd02c6d1435> in <module>
----> 1 f(0.1)

<lambdifygenerated-1> in _lambdifygenerated(x)
      2     return (  # Not supported in Python with SciPy:
      3   # Contains
----> 4 logical_and.reduce((greater(x, 0),Contains(x, Interval(-1, 1)))))

NameError: name 'Contains' is not defined
```

**Comment by craymichael** [comment]
Apologies - I should have made this more clear. Firstmost, here is the 1.6 code that does indeed work:

```python
>>> from sympy.stats.rv import sample_iter_subs
>>> next(sample_iter_subs(U, interval.contains(U)))
0.0902218444596365
```

The old code for `sample_iter` looks like this:

```python
    try:
        return sample_iter_lambdify(expr, condition, size=size, numsamples=numsamples, **kwargs)
    # use subs when lambdify fails
    except TypeError:
        return sample_iter_subs(expr, condition, size=size, numsamples=numsamples, **kwargs)
```

In the case that lambdify fails, it falls back on using `xreplace` in evaluation of conditions. However, the old code only assumed a `TypeError` could be raised as shown above. The 1.7 code performs the same operation for both `sample_iter_subs` and `sample_iter_lambdify`. Does it make sense to call this a regression? This is an expected failing of lambdify and I would consider it more an issue of how the condition is evaluated

**Comment by czgdp1807** [comment]
AFAICT, the `1.7` code introduced consistency for all sampling APIs and that is why same operation was performed for the calls you said. It was intended actually. I can create a work around for stats by applying a try-except whenever the lambdify fails but still that won't fix lambdify, it will just bypass the failure. 
P. S. I will be looking at in more detail by tomorrow.

**Comment by craymichael** [comment]
That makes sense. I believe that consistency would be maintained if the try-except were added around the condition function `given_fn`.
```
     if condition is not None:
        # TODO add try-except for TypeError and NameError here
        given_fn = lambdify(rvs, condition, **kwargs)
```
It would slow things down but increase functionality. As for fixing lambdify I'm much less familiar with but look forward to hear back when you have the time later. Appreciate the quick responses here!

**Comment by oscarbenjamin** [comment]
This is with sympy 1.7 and seems to work:
```
In [1]: from sympy.stats.rv import sample_iter_subs                                                                               

In [2]: from sympy.stats import *                                                                                                 

In [3]: U = Uniform('U', -1, 1)                                                                                                   

In [4]: next(sample_iter_subs(U, Interval(-1, 1).contains(U)))                                                                    
Out[4]: 0.6434044191802784
```
Am I misunderstanding something?

Can you show a complete code sample with all imports etc that illustrates the difference between 1.6.2 and 1.7?

**Comment by craymichael** [comment]
Yes, the problem becomes when the `Contains` expression does not evaluate prior to sampling. Your interval is evaluated immediately to something that can be lambdified.

In detail, my provided case:

```python
>>> interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
>>> interval
ConditionSet(x, x > 0, Interval(-1, 1))
>>> interval.contains(U)
(U > 0) & Contains(U, Interval(-1, 1))
```

Your case:

```python
>>> interval = S.Interval(-1, 1)
>>> interval
Interval(-1, 1)
>>> interval.contains(U)
(-1 <= U) & (U <= 1)
```

**Comment by craymichael** [comment]
Which imo is strange behavior as further simplification of Contains should be possible...but perhaps not in cases with more complex ConditionSets? I first discovered this bug with a fairly complicated function

**Comment by czgdp1807** [comment]
IMHO, this may not be something that can be called regression as the behavior of public API (sample) is still the same across `1.6` and `1.7`(`sample_iter_subs` and `sample_iter_lambdify` are not in the public APIs). In fact both these functions now delegate the task to `sample_iter` (which is public). 
Though this is still an issue. A good and permanent fix for `stats` module would be to change the way it verifies the satisfaction of given condition.

**Comment by oscarbenjamin** [comment]
I feel like I'm being stupid here :)

Can you please help me out by showing a *complete* example of the code that shows the difference?

I tried this
```python
import sympy as S
from sympy.stats import *

x = S.Symbol('x')
U = Uniform('U', -1, 1)
interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
print(interval.contains(U))
```
I get the same output with both sympy 1.6.2 and 1.7:
```console
$ python er.py 
(U > 0) & Contains(U, Interval(-1, 1))
```

**Comment by craymichael** [comment]
Sure thing - here is something that works in 1.6

```python
>>> import sympy as S
>>> S.__version__
'1.6'
>>> from sympy import stats
>>> from sympy.stats.rv import sample_iter_subs
>>> x = S.Symbol('x')
>>> interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
>>> U = stats.Uniform('U', -1, 1)
>>> next(sample_iter_subs(U, interval.contains(U)))
0.210479150398788
```

That fails in 1.7

```python
>>> import sympy as S
>>> S.__version__
'1.7'
>>> from sympy import stats
>>> from sympy.stats.rv import sample_iter_subs
>>> x = S.Symbol('x')
>>> interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
>>> U = stats.Uniform('U', -1, 1)
>>> next(sample_iter_subs(U, interval.contains(U)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/site-packages/sympy/stats/rv.py", line 1220, in return_generator
    gd = given_fn(*args)
  File "<lambdifygenerated-2>", line 4, in _lambdifygenerated
NameError: name 'Contains' is not defined
```

**Comment by czgdp1807** [comment]
Can you please show the differences in public APIs? The ones in your example aren't public.

**Comment by craymichael** [comment]
Ah, they'll fail the same way in the public API in both 1.6 and 1.7 - there is no difference in `sample_iter`. However, it would be nice if being able to conditionally sample this way were supported. The 1.6 function doc for `sample_iter` says
```
    See Also
    ========

    sample
    sampling_P
    sampling_E
    sample_iter_lambdify
    sample_iter_subs
```
so it wasn't entirely clear this wasn't in the public API (other than per the full import). Perhaps that's me being naive

Would cooking up a failure between 1.6 and 1.7 public API here aid progress in this issue? Or is the case included enough to spark changes? I can help contribute to a fix as well

**Comment by czgdp1807** [comment]
Sure. Since now we are not in a hurry, may be we can discuss more on how to approach this problem. Should we add a work around in stats module or should we modify lambdify to work with `Contains`?

**Comment by oscarbenjamin** [comment]
> so it wasn't entirely clear this wasn't in the public API (other than per the full import). Perhaps that's me being naive

I don't think the docs are very clear about what is public and what is not. I'm personally not very familiar with e.g. the sampling code and I have no idea what is supposed to be private...

**Comment by craymichael** [comment]
That is true. Perhaps this issue could even be addressed in how simplification occurs? For instance, take a look at the example:

```python
>>> U = stats.Uniform('U', -1, 1)
>>> interval = S.ConditionSet(x, S.Gt(x, 0), S.Interval(-1, 1))
>>> interval
ConditionSet(x, x > 0, Interval(-1, 1))
>>> interval.contains(U)
(U > 0) & Contains(U, Interval(-1, 1))
```
Like shouldn't this be able to simplify to at least `(-1 <= U) & (U <= 1) & (U > 0)` (equivalent to `(U <= 1) & (U > 0)`)? Lambdify could handle that case

Lambdify would also have to work with `Interval` (as that is also not evaluated). Both should be fairly straightforward (I think) to implement efficiently.

**Comment by czgdp1807** [comment]
I agree with @oscarbenjamin about the docs part. Actually they need to be updated and the content is required to be kept up to date. I opened an issue some time back for the same. Will try to find some time during the vacations to work on the docs. 
Regarding the fix, I would suggest to modify `lambdify` as it will help code other than `stats` module as well. Adding workarounds to stats module will be more of ad-hoc fix and wouldn't last long before another issue arises.

**Comment by oscarbenjamin** [comment]
To be clear this is the what fails in 1.7 although it passes in 1.6.2:
```python
from sympy import *
from sympy.stats import Uniform
from sympy.stats.rv import sample_iter_subs

U = Uniform('U', -1, 1)
next(sample_iter_subs(U, (U > 0) & Contains(U, Interval(-1, 1)))) 
```
Simplification of Contains etc can be improved but is not what has *changed*. The difference is to do with `sample_iter_subs` somehow. I don't know how that function has changed but @czgdp1807 says that it is a "private" API although it isn't clearly documented as such.

I have a few questions:

1. What exactly has changed?
2. Is there a recommended way that is better (or a workaround)?
3. Is this easy or worthwhile to fix so that we could just fix it in 1.7.1?

**Comment by craymichael** [comment]
1. The behavior of `sample_iter` has changed between the two versions.
    i. In [1.6.2](https://github.com/sympy/sympy/blob/sympy-1.6.2/sympy/stats/rv.py#L1078), `sample_iter` attempts to use `sample_iter_lambdify`. If it fails due to `TypeError`, it then tries to use `sample_iter_subs` instead.
    ii. In [1.7](https://github.com/sympy/sympy/blob/sympy-1.7/sympy/stats/rv.py#L1185), lambdify is used exclusively. Furthermore, both `sample_iter_lambdify` and `sample_iter_subs` were deprecated. I believe this all was done for consistent sampling using external libraries, see https://github.com/sympy/sympy/issues/19061 and https://github.com/sympy/sympy/pull/18754/files
2. I mentioned a simple workaround (that wouldn't impact sampling consistency) earlier in this issue https://github.com/sympy/sympy/issues/20563#issuecomment-741973209 - however, we discussed instead fixing the issue in lambdify
3. I am not familiar enough with the sympy codebase to give you a definitive answer.

Let me know if that clears things up

**Comment by oscarbenjamin** [comment]
> 2\. I mentioned a simple workaround

I meant "workaround" in the sense that users could change *their* code to work around the issue rather than the code in *sympy* being changed.

**Comment by craymichael** [comment]
Ah, got it. One hacky way I can think of is to instead have the condition in their own code, filtering out bad samples until they get the number of desired samples. It'll definitely be slow.

```python
from sympy import *
from sympy.stats import Uniform
from sympy.stats import sample_iter

desired_samples = 10
U = Uniform('U', -1, 1)
condition = (U > 0) & Contains(U, Interval(-1, 1))

samples = []
for x in sample_iter(U):
    if condition.subs({U: x}):
        samples.append(x)
        if len(samples) == desired_samples:
            break
```

**Comment by czgdp1807** [comment]
Following is an ad-hoc fix,
```python
diff --git a/sympy/stats/rv.py b/sympy/stats/rv.py
index a982137c9c..6239fb9867 100644
--- a/sympy/stats/rv.py
+++ b/sympy/stats/rv.py
@@ -1225,12 +1225,21 @@ def sample_iter(expr, condition=None, size=(), library='scipy',
                 sub[arg] = RandomSymbol(arg.symbol, arg.pspace)
         expr = expr.subs(sub)
 
+    def fn_subs(*args):
+        return expr.subs({rv: arg for rv, arg in zip(rvs, args)})
+
+    def given_fn_subs(*args):
+        if condition is not None:
+            return condition.subs({rv: arg for rv, arg in zip(rvs, args)})
+        return False
+
     if library == 'pymc3':
         # Currently unable to lambdify in pymc3
         # TODO : Remove 'pymc3' when lambdify accepts 'pymc3' as module
         fn = lambdify(rvs, expr, **kwargs)
     else:
         fn = lambdify(rvs, expr, modules=library, **kwargs)
+
     if condition is not None:
         given_fn = lambdify(rvs, condition, **kwargs)
 
@@ -1246,7 +1255,10 @@ def return_generator_infinite():
             args = [d[rv] for rv in rvs]
 
             if condition is not None:  # Check that these values satisfy the condition
-                gd = given_fn(*args)
+                try:
+                    gd = given_fn(*args)
+                except (NameError, TypeError):
+                    gd = given_fn_subs(*args)
                 if gd != True and gd != False:
                     raise ValueError(
                         "Conditions must not contain free symbols")
@@ -1267,7 +1279,10 @@ def return_generator_finite():
                 args = [d[rv][count] for rv in rvs]
 
                 if condition is not None:  # Check that these values satisfy the condition
-                    gd = given_fn(*args)
+                    try:
+                        gd = given_fn(*args)
+                    except (NameError, TypeError):
+                        gd = given_fn_subs(*args)
                     if gd != True and gd != False:
                         raise ValueError(
                             "Conditions must not contain free symbols")
@@ -1279,7 +1294,10 @@ def return_generator_finite():
         count = 0
         while count < numsamples:
             args = [d[rv][count] for rv in rvs]
-            yield fn(*args)
+            try:
+                yield fn(*args)
+            except (NameError, TypeError):
+                yield fn_subs(*args)
             count += 1
 
     if numsamples is S.Infinity:
diff --git a/sympy/stats/tests/test_rv.py b/sympy/stats/tests/test_rv.py
index 9027b0e00b..02446cc36f 100644
--- a/sympy/stats/tests/test_rv.py
+++ b/sympy/stats/tests/test_rv.py
@@ -1,11 +1,13 @@
 from sympy import (S, Symbol, Interval, exp, Or,
         symbols, Eq, cos, And, Tuple, integrate, oo, sin, Sum, Basic, Indexed,
-        DiracDelta, Lambda, log, pi, FallingFactorial, Rational, Matrix)
+        DiracDelta, Lambda, log, pi, FallingFactorial, Rational, Matrix,
+        ConditionSet, Gt)
 from sympy.stats import (Die, Normal, Exponential, FiniteRV, P, E, H, variance,
         density, given, independent, dependent, where, pspace, GaussianUnitaryEnsemble,
         random_symbols, sample, Geometric, factorial_moment, Binomial, Hypergeometric,
         DiscreteUniform, Poisson, characteristic_function, moment_generating_function,
-        BernoulliProcess, Variance, Expectation, Probability, Covariance, covariance)
+        BernoulliProcess, Variance, Expectation, Probability, Covariance, covariance,
+        Uniform)
 from sympy.stats.rv import (IndependentProductPSpace, rs_swap, Density, NamedArgsMixin,
         RandomSymbol, sample_iter, PSpace, is_random, RandomIndexedSymbol, RandomMatrixSymbol)
 from sympy.testing.pytest import raises, skip, XFAIL, ignore_warnings
@@ -116,6 +118,12 @@ def is_iterator(obj):
     for _ in range(10):
         assert next(itr_inf_1) == next(itr_inf_2)
 
+    x = Symbol('x')
+    interval = ConditionSet(x, Gt(x, 0), Interval(-1, 1))
+    U = Uniform('U', -1, 1)
+    s1 = sample(U, interval.contains(U), seed=0)
+    assert interval.contains(next(s1))
+
 def test_pspace():
     X, Y = Normal('X', 0, 1), Normal('Y', 0, 1)
     x = Symbol('x')
```
Let me know where this should go to, `1.7` or to `master`?

**Comment by czgdp1807** [comment]
I see, `sample_iter_subs`, `sample_iter_lambdify` aren't used anywhere and neither are the accessible through `sympy.stats`. So, I think we should remove these two functions in `1.8`. In addition, I think we should decide on one, either `lamdify` or `subs`. Using both seems heuristic and quite un-predictable. What do you say @Upabjojr ?

**Comment by craymichael** [comment]
Butting in here - `sample_iter_lambdify` and `sample_iter_subs` have been [deprecated](https://github.com/sympy/sympy/pull/18754/commits/c9ede553f839ab39de7dd27da94873db33ca7dcd). In my experience `lambdify` is significantly faster than `subs`, especially with various supported backends. Both functions would also use the same RNG in the backend, so it would yield the same results (especially with the 1.7.1-introduced seed in `sample` https://github.com/sympy/sympy/pull/20528/files), no?

**Comment by czgdp1807** [comment]
> Butting in here - `sample_iter_lambdify` and `sample_iter_subs` have been [deprecated](https://github.com/sympy/sympy/pull/18754/commits/c9ede553f839ab39de7dd27da94873db33ca7dcd).

Actually, we were going to deprecate the above functions but didn't. I also made a suggestion to remove the deprecation which was applied later on.


>  In my experience `lambdify` is significantly faster than `subs`, especially with various supported backends.

Well, that is why I am not up for any ad-hoc fixes to just `sympy.stats` for cases where `lamdify` fails. If `lamdify` is efficient then probably, we should fix it rather than going for ad-hoc workarounds as they won't last long. Though, if a quick fix is required for `1.7.1` then I am ready to submit a PR for it. The reason why I am reluctant to do it is that the APIs which differ in behavior between `1.6` and `1.7` are not supposed to be used  by users. They are internal (i.e., not accessible via `sympy.stats` and one would need to use `sympy.stats.rv`).  So, if one can fix `lamdify` for unevaluated objects like, `Contains`, `Interval` then that would be more beneficial and would be a lasting fix.

**Comment by craymichael** [comment]
Ah, I see now. Thank you for the information. I agree that integrating with `lambdify` makes more sense here, but this would be sufficient in the short-term

**Comment by czgdp1807** [comment]
Okay. I would submit a PR to `1.7` after #20555 is merged, though would a add `#TODO` to remove the ad-hoc fix once `lambdify` is fixed in future.

**Comment by craymichael** [comment]
Ok, so I actually just made a potential fix to `lambdify` - my test cases (not thrown in yet) are below.

```python
from sympy import Interval, Contains
from sympy.stats import Uniform, sample_iter
from sympy.printing.pycode import PythonCodePrinter, NumPyPrinter, SciPyPrinter, MpmathPrinter, SymPyPrinter
from sympy.printing.lambdarepr import NumExprPrinter
from sympy.printing.tensorflow import TensorflowPrinter

U = Uniform('U', -1, 1)
expr = (U > 0) & Contains(U, Interval(-1, 1))

for printer in (PythonCodePrinter, NumPyPrinter, SciPyPrinter, MpmathPrinter, NumExprPrinter, TensorflowPrinter):
    code = printer().doprint(expr)
    print(printer.__name__ + '\n   ', code)

for module in ('numpy', 'scipy', 'pymc3'):
    print(module + '\n   ', next(sample_iter(U, (U > 0) & Contains(U, Interval(-1, 1)), library=module)))
```
The output

```
PythonCodePrinter
    (-1 <= U) and (U <= 1) and (U > 0)
NumPyPrinter
    numpy.logical_and.reduce((numpy.greater(U, 0),numpy.logical_and.reduce((numpy.less_equal(-1, U),numpy.less_equal(U, 1)))))
SciPyPrinter
    numpy.logical_and.reduce((numpy.greater(U, 0),numpy.logical_and.reduce((numpy.less_equal(-1, U),numpy.less_equal(U, 1)))))
MpmathPrinter
    (-1 <= U) and (U <= 1) and (U > 0)
NumExprPrinter
    evaluate('(((((-1 <= U)) and ((U <= 1)))) and ((U > 0)))', truediv=True)
TensorflowPrinter
    tensorflow.math.logical_and(tensorflow.math.greater(U, 0), tensorflow.math.logical_and(tensorflow.math.less_equal(-1, U), tensorflow.math.less_equal(U, 1)))
numpy
    0.8282566197978278
scipy
    0.2525007729573858
```

Strangely `pymc3` gives me an exception (due to size being a tuple and directly passed to `pymc3` sampler with expects an int) (working of sympy master last commit `e5c901e5f30bcc9bd0cb91f03d0696815b96bd1a` and `pymc3` version `3.10.0`)


```
Traceback (most recent call last):
  File "/usr/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 15, in <module>
  File "/mnt/media/Projects/sympy/sympy/stats/rv.py", line 1216, in return_generator
    d = ps.sample(size=size, library=library)  # a dictionary that maps RVs to values
  File "/mnt/media/Projects/sympy/sympy/stats/crv.py", line 644, in sample
    return {self.value: self.distribution.sample(size, library=library)}
  File "/mnt/media/Projects/sympy/sympy/stats/crv.py", line 333, in sample
    samps = _get_sample_class_crv[library](self, size)
  File "/mnt/media/Projects/sympy/sympy/stats/crv.py", line 251, in __new__
    return cls._sample_pymc3(dist, size)
  File "/mnt/media/Projects/sympy/sympy/stats/crv.py", line 288, in _sample_pymc3
    return pymc3.sample(size, chains=1, progressbar=False)[:]['X']
  File "/mnt/media/Projects/sympy/venv/lib/python3.8/site-packages/pymc3/sampling.py", line 481, in sample
    elif draws < 500:
TypeError: '<' not supported between instances of 'tuple' and 'int'
```

The proposed patch:
```python
diff --git a/sympy/printing/pycode.py b/sympy/printing/pycode.py
index b7315a4c55..79b8bae562 100644
--- a/sympy/printing/pycode.py
+++ b/sympy/printing/pycode.py
@@ -425,6 +425,11 @@ def _hprint_Pow(self, expr, rational=False, sqrt='math.sqrt'):
         exp_str = self.parenthesize(expr.exp, PREC, strict=False)
         return "{}**{}".format(base_str, exp_str)
 
+    def _print_Contains(self, expr):
+        """Converts a `Contains()` object into relational expressions"""
+        value, set_ = expr.args
+        return self._print(set_.as_relational(value))
+
 
 class PythonCodePrinter(AbstractPythonCodePrinter):
```

**Comment by oscarbenjamin** [comment]
That seems like an improvement but it's not always the case that Contains can be rewritten with `as_relational` which is mostly only for subsets of the real line e.g.:
```
In [8]: c = Contains(x, Complexes)                                                                                                

In [9]: c                                                                                                                         
Out[9]: x ∈ ℂ

In [10]: c.args[1].as_relational(x)                                                                                               
---------------------------------------------------------------------------
AttributeError
```
I'm not sure that an arbitrary `Contains` can be lambdified but maybe it's worthwhile to try something that at least works most of the time.

What exactly is the purpose of the condition in sample? I was surprised to see that it is possible to pass in a Contains like that.

**Comment by craymichael** [comment]
Hmm...good point. I believe that anything that has `as_relational` would work though. So I ran a quick filter of the public API and found the following don't have the attribute-

```python
import sympy
from sympy.sets import sets
all_objects = [getattr(sympy, s) for s in dir(sympy)]
mros = [(getattr(s, '__mro__', getattr(s.__class__, '__mro__'))) for s in all_objects]  # why are some of these a bool
print([s[0].__name__ if hasattr(s[0], '__name__') else s[0] for s in filter(lambda om: (isinstance(om[0], sets.Set) or (type(om[1]) is not bool and sets.Set in om[1])) and not hasattr(om[0], 'as_relational'), zip(all_objects, mros))])
```

This gives
```python
['Circle', 'ComplexRegion', S.Complexes, 'Curve', 'DisjointUnion', 'Ellipse', 'ImageSet', 'Line', 'Line2D', 'Line3D', 'Parabola', 'Polygon', 'PowerSet', Rationals, 'Ray', 'Ray2D', 'Ray3D', 'RegularPolygon', 'Segment', 'Segment2D', 'Segment3D', 'Set', 'Triangle']
```

I think that, eventually, `as_relational` support to each non-abstract `Set` object could be added  as it has good mappings to operators available in all backends, and for now raise a `NotImplementedError` if a `Set` doesn't have this attribute? This would satisfy a large amount of cases, and `as_relational` could definitely be added for most of the above sets (I think)

So, the context where I found the conditions to be useful was in a project I've been working on. Within, I automatically generate a bunch of crazy random expressions in some of the research I'm working on, then discover the valid domain of real numbers that I can draw samples from (`continuous_domain` for each free symbol). I constrain those samples to something like Uniform(-1, +1), then use the additionally found constraints (valid real intervals) to filter out invalid samples. I was also surprised that Contains worked but in this sampling context I think that it makes sense

**Comment by oscarbenjamin** [comment]
> `as_relational` could definitely be added for most of the above sets (I think)

What would `as_relational` be for any of those examples? It's a method that takes only one symbol so anything more than 1D doesn't really work.

> I found the conditions to be useful was in a project I've been working on

Thanks for the explanation. It's always good to hear what sympy is being used for and good to know that it's useful!

**Comment by craymichael** [comment]
Oh that's a good point - I'm now seeing I didn't think this through enough, haha. I think N-D would have to be a function of the sampling function parameter `size` - like each sample would be size `(3,)` in `sample_iter` for a `Line3D` or `(2,)` for a `Circle`. Then whether they fall on the line(s) would dictate a valid sample. However, this would be incredibly constraining on the sample function which does a filter - sampling from the line would be far more efficient.

For several of these I think special cases would need to be involved. A `Contains` of a `Union` however should be possible.

> Thanks for the explanation

No problem! This library has been super fun to work with and I'm doing my best to push it to some of its limits!

**Comment by czgdp1807** [comment]
I have opened #20580 , it is not necessary to merge it as it's just a quick fix workaround for stats. The issue with `lambdify` will still persist.

**Comment by oscarbenjamin** [comment]
We have this:
```
In [14]: Reals.as_relational(x)                                                                                                   
Out[14]: -∞ < x ∧ x < ∞
```
So maybe for Complexes it could be `Abs(x) < oo`. Unfortunately that just evaluates to True (which is incorrect if x is not declared as finite):
```
In [1]: x = Symbol('x')                                                                                                           

In [2]: Abs(x) < oo                                                                                                               
Out[2]: True

In [3]: x = Symbol('x', finite=False)                                                                                             

In [4]: Abs(x) < oo                                                                                                               
Out[4]: True
```

**Comment by craymichael** [comment]
I think that makes sense and I would call `Abs(x) < oo` evaluating to `True` when `x` is not finite a bug. We could make this work as follows for now-

```python
>>> from sympy import printing
>>> expr = Lt(Abs(x), oo, evaluate=False)
>>> expr
Abs(x) < oo
>>> printing.pycode(expr)
"(abs(x) < float('inf'))"
```

Furthermore, I find this bit strange...
```python
>>> x = Symbol('x', infinite=True)
>>> Abs(x) < oo
True
>>> x = Symbol('x', infinite=True, finite=False)
>>> Abs(x) < oo
True
```
This is a bug then, no? Or do I misunderstand assumptions?

**Comment by oscarbenjamin** [comment]
> This is a bug then, no? Or do I misunderstand assumptions?

You're correct. It is a bug. I'm not sure where that happens.

**Comment by craymichael** [comment]
So I'm thinking for `Complexes` we have:

```python
    def as_relational(self, x):
        """Rewrite a Range in terms of equalities and logic operators. """
        from sympy.functions.elementary.complexes import Abs
        abs_x = Abs(x)
        return And(Lt(-oo, abs_x, evaluate=False),
                   Lt(abs_x, oo, evaluate=False))
```

For `DisjointUnion` we have:
```python
    def as_relational(self, x):
        """Rewrite a Range in terms of equalities and logic operators. """
        return self.rewrite(Union).as_relational(x)
```

This leaves the following:

2+D Sets:

- Circle
- Curve
- Ellipse
- Line (Line2D, Line3D)
- Parabola
- Polygon (RegularPolygon)
- Ray (Ray2D, Ray3D)
- Segment (Segment2D, Segment3D)
- Triangle

Trickier 1D Sets (2D for complex I guess)

- ComplexRegion (takes complex numbers or polar coordinates as arguments)
- PowerSet
- Rationals (no idea if this is possible unless input types restricted to be rationals/ints)

Others:

- ImageSet
- Set (not meant to be used directly so we can leave not implemented)

I think that what we have now should cover most practical use cases involving lambdify?

**Comment by oscarbenjamin** [comment]
Do you mean the `is_ge` function somewhere?

**Comment by craymichael** [comment]
Yes I mean that. But also that's not the only problem so I'm not entirely certain what's up. I just opened https://github.com/sympy/sympy/issues/20584 with another example where `Ge` fails
